### ISSUE TITLE
Thread Merge 

### ISSUE THREAD

**Issue by wjbbupt** [eclipse-cyclonedds/cyclonedds#1710]
Hello, I would like to do the following things,

1. Merge three receiving threads, including recv recuMU, into one receiving thread,

2. dq. builtin thread removal

3. The goal is to leave three threads of tev\gc \recv. Is it reasonable to modify this idea under the current framework, and can it be modified directly

**Comment by eboasson** [comment]
Hi @wjbbupt, it is not going to be *that* simple 🙂:

There is a setting to configure a single or multiple receive threads: `Internal/MultipleReceiveThreads`. Removing the ability to have multiple receive threads is not something I am willing to contemplate, but changing the default might make sense. That will add some CPU usage and latency, but for most applications that won't matter and it would eliminate the dreaded "Cyclone hangs on shutdown". So that's definitely worth considering.

The `dq.builtin` thread is quite important. It does all the processing for incoming discovery data, and the impact on the application of processing those synchronously on the receive thread can be quite bad. Whether that impact is a problem is dependent on the application, and I am willing to believe there are systems for which this would be an acceptable trade-off. From that perspective, adding a setting to do the processing synchronously could be an option.

Do note that doing this processing asynchronously is a choice that dates back a very long time, and changing long-standing assumptions on how things work usually brings risk.

Starting the **`dq.user`** thread only when needed (thinking of #1709 here, too) is an interesting option. That should be fairly straightforward, but since it is simply sleeping about 100% of the time I never really bothered looking into it. On small RTOS it would definitely be good.

**Comment by linghunguntang** [comment]
Hi@eboasson ，Regarding your response to the merge of recv threads, modifying the default value is effective.

I would like to ask you again why arg.mode in the "recv" thread is set to DDSI_RTMMANY, but recvUC and recvMC are set to DDSI_RTM_SINGLE. After closing recvUC and recvMC, only the recv thread remains, and ddsi_sock_waitset must also be used? Please further explain the design concept of receiving data threads.

**Comment by eboasson** [comment]
> I would like to ask you again why arg.mode in the "recv" thread is set to DDSI_RTMMANY, but recvUC and recvMC are set to DDSI_RTM_SINGLE. After closing recvUC and recvMC, only the recv thread remains, and ddsi_sock_waitset must also be used? Please further explain the design concept of receiving data threads.

Yes, it is simply whether the thread is handling only one socket or (potentially) multiple. If there is only one it can use a blocking `recvfrom` which is a bit faster than doing the `select`+`recvfrom` (or a variant thereof) for multiplexing I/O on one thread. It is not really a meaningful difference in most systems, but sometimes it is.

So there is that, and it can do most of the processing in parallel if there is a lot of data arriving on different sockets. That's not fully exploited today, but the DDSI protocol and most of the implementation allow for creating sockets for, e.g., individual topics, and then that becomes interesting. The restrictions in the implementation are really about configuring it, there are just no options for it at the moment.

**Comment by linghunguntang** [comment]
Hi @eboasson ，
In ddsi_ In the socketwaitset. c file, the function ddsi_sock_waitset_next_event 
`unsigned idx = ctx->index++` And ddsi_sock_waitset_wait `/* this simply skips the trigger fd */
    ctx->index = 1;` is the operation repeated, causing the descriptor corresponding to conn [1] to never be FD_ISSET

**Comment by linghunguntang** [comment]
Hi @eboasson ：
I would like to ask about the design concept of creating pipelines in the function `ddsi_sock_waitset_new`, as well as the application scenarios of the function `ddsi_sock_waitset_trigger`. What side effects will occur if manual triggering is not performed

**Comment by eboasson** [comment]
The pipe exists solely for the purpose of ensuring it wakes up reasonably quickly when:
* terminating, so that the absence of network traffic doesn't mean it hangs;
* when adding a new socket to the set of sockets to read from, which for UDP happens when you create a participant and `ManySocketsMode` is true, and for TCP happens whenever a new connection is created;
* when it is made "deaf" (a testing trick) because of at some point it would actually delete the sockets and leave multicast groups — but I suspect it can be deleted now.

So when using UDP it is really no worse than hanging on termination. Simply adding a time out to the wait also works :)
